### ISSUE TITLE
SUGGESTION: Add methods to drop instances (databases) and stores (collections)

### ISSUE THREAD

**Issue by jlchereau** [localForage/localForage#620]
Is this hidden somewhere and have I missed it?
Dropping/deleting collections would be really useful, the database is less urgent.

**Comment by thgreasi** [comment]
Is [`clear()`](http://localforage.github.io/localForage/#data-api-clear) too much for your needs?
Does something like [removeItems](https://github.com/localForage/localForage-removeItems) sound better?

**Comment by jlchereau** [comment]
There is a difference between ```DROP DATABASE``, ```DROP TABLE``` and ```DELETE * FROM TABLE``` in SQL. 
If SQL makes this difference, if mongoDB and others make it too, there might be a need for it in localForage, don't you think?

My need is to upgrade the schema without leaving empty useless stores/tables/collections after the upgrade.  Does it  make sense?

**Comment by thgreasi** [comment]
I was 100% with you about the DROP part, I was just asking about the collection data part that you mentioned.
I will do my best to add such a method in v1.5.
I would like to hear some api suggestions. Maybe something like dropInstance() or clearInstance() or deleteInstance(). I guess that we should be able to invoke it:
* without arguments to drop the table of the current instance
* with 2 string arguments defining the table of the db that we need to drop
* with 1 string argument defining the db we need to drop
Any thoughts on these? (cc: @tofumatt)

An extra consideration is whether we should add the new method as a required or optional method of the custom driver api. Making it required would make the new release backwards incompatible.

**Comment by jlchereau** [comment]
I have only ever experienced different flavors of SQL databases and mongoDB, and they both *drop* so I would definitely favor **dropInstance** over clearInstance and deleteInstance. Furthermore clearInstance is ambiguous considering the clear method in localForage.

Considering the existing createInstance:

```
localforage.createInstance({ name: dbName, storeName: storeName });
```

It makes sense to maintain the same method signature (as you suggest):

```
localforage.dropInstance({ name: dbName, storeName: storeName });
```

As far as backward compatibility is concerned, I have no idea how many people have developed drivers for localForage and how easy it is to communicate with them. In my opinion, backward incompatibility is not a big issue if localForage can report its version like jQuery has ```$.fn.jquery```.

**Comment by thgreasi** [comment]
@jlchereau OK, `dropInstance` was all I needed to hear. I will try to take a look at it during the weekened.

**Comment by thgreasi** [comment]
Still working on this. My initial findings include:
* **IndexedDB** in order to drop a "table", all the connection to the same storage need to close and the version must increase. This might cause issues for people that are creating instances in their code with predefined versions. Finding a way to reopen all the db connections on all related instances might complicate the driver. I think that, if we are going to make this method available, we should do that (reopen the dbs), what's your opinion?
* **WebSQL** have written the code, but still haven't done any testing with it.
* ***localstorage** we just have to delete the records with the respective key prefixes for the requested "db" name and "table" name. This revives the issue that we still have (I think) where `clear()` the default "table" of a "db" actually `clear()`s all the db. Can someone confirm this?

**Comment by jlchereau** [comment]
- **IndexedDB**: agreed.
- **WebSQL**: n/a
- **localStorage**: I would need to try to confirm (after Christmas).

**Comment by thgreasi** [comment]
Changing this to milestone v1.6 since it currently seems that it needs more work.

**Comment by thgreasi** [comment]
For reference, the current status on this can be reviewed on [thgreasi/localForage/tree/dropInstance](https://github.com/thgreasi/localForage/tree/dropInstance).

**Comment by thgreasi** [comment]
I've pushed a working branch on https://github.com/thgreasi/localForage/tree/drop-instance-v2 . It would be great if anyone could do some extra testing on this. I'm super excited about this starting to work and have already added some tests.

**Comment by thgreasi** [comment]
@jlchereau The `dropInstance()` PR in in the current master (and planned to be released as LF v1.6) can you give it a try?

**Comment by jlchereau** [comment]
Thanks you for progressing this. At this stage, I have only tried with Chrome 62 on Windows 10, which uses indexedDB by default.

localforage.dropInstance({ name: 'db', storeName: 'table' }, function(err) {...}); works fine when both db and table exist and err is null which is consistent with all localForage methods.

localforage.dropInstance({ name: 'db' }, function(err) {...}); returns err of type IDBVersionChangeEvent and the database db is not removed which is unexpected.

localforage.dropInstance({ name: 'dummy', storeName: 'dummy' }, function(err) {...}); where dummy does not exist raises Uncaught TypeError: Cannot read property 'deferredOperations' of undefined which could be handled more nicely.

I hope this helps.

**Comment by jlchereau** [comment]
Also, any reason why dropping the first instance is 10x longer than dropping the second one? Maybe you want to look into that too...

![image](https://user-images.githubusercontent.com/2556751/32487006-425c8282-c3a0-11e7-967c-67f3f9390b57.png)

**Comment by sido420** [comment]
Any updates on this?

**Comment by thgreasi** [comment]
@sido420 are interested on a specific part?
`dropInstance` is available as part of LF v1.6 and thus this issue was closed.
